### PR TITLE
add support for slicing, subsampling and splitting StreamingDataset

### LINKED ISSUE #135: Subsample StreamingDataset
## 🚀 Feature
`StreamingDataset` has enabled fast data reading, which is amazing when we have a large dataset. However, currently, it does not support reading just a fraction of data, and simple methods such as slicing the dataset does not work with `StreamingDataset`.

Could we have a feature to support this? Thanks. 

cc: @tchaton

### LINKED ISSUE #145: Support splitting datasets
## 🚀 Feature

Support splitting of datasets

### Motivation

My dataset doesn't have train/val/test split and I get `RuntimeError: The provided dataset should be either an instance of StreamingDataset or CombinedStreamingDataset. Found <torch.utils.data.dataset.Subset object>` when using random splitting. I would like this workflow supported

### Pitch

The approach above is compatible with streaming datasets

### Alternatives

Pre splitting the dataset

### Additional context

- A StreamingDataset is described by its index.json. The chunks are loaded from the file into a list there: https://github.com/Lightning-AI/litdata/blob/main/src/litdata/streaming/config.py#L56
[3:58](https://pytorch-lightning.slack.com/archives/D05N141TKF0/p1716994719839439)
- We could have a utility that takes a StreamingDataset and return multiple ones with sliced chunks list.
- Overlaps https://github.com/Lightning-AI/litdata/issues/135

### PR DESCRIPTION
**Pull request by deependujha**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/agreed via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lit-data/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes #135 & fixes #145.

1. Adds support to **slice StreamingDataset**
Screenshot from 2024-06-05 13-20-06

2. Adds support to **Subsample StreamingDataset** 
Screenshot from 2024-06-08 01-29-54

3. Adds support to **train_test_split StreamingDataset** 
Screenshot from 2024-06-08 01-29-18

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in GitHub issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by tchaton** [comment]
Hey @deependujha, how is it going ?

**Comment by deependujha** [comment]
I modified `item_loader interval`, and instead of returning a list of `[start_chunk_idx, end_chunk_idx]`, it returns `[start_chunk_idx, my_chunk_start, my_chunk_end, end_chunk]`.

Refer to this line: `streaming/item_loader#192`.

`my_chunk_start` and `my_chunk_end` denotes from which index to which index of the current chunk, this streaming dataset is allowed to read.

If we are to read the whole chunk, it will simply be `[start_chunk_idx, start_chunk_idx,  end_chunk_idx, end_chunk_idx]`.

The logic for subsampling:
Screenshot from 2024-06-07 19-21-23

The logic for train_test_split
Screenshot from 2024-06-07 19-26-05

---

Using each chunk as a basis for subsampling and train_test_split.

File: src/litdata/streaming/dataset.py [hunk 1]
```diff
@@ -297,7 +317,7 @@ def __next__(self) -> Any:
             self.index = 0
 
             interval = self.worker_intervals[self.chunk_index]
-            current_indexes = np.arange(interval[0], interval[1])
+            current_indexes = np.arange(interval[1], interval[2])
```
Comment (tchaton) [inline src/litdata/streaming/dataset.py:320]: why this change ?

File: src/litdata/streaming/config.py [hunk 2]
```diff
@@ -58,7 +60,7 @@ def __init__(
         self._config["data_spec"] = treespec_loads(self._config["data_spec"])
 
         self._item_loader.setup(self._config, self._chunks, serializers)
-        self._intervals = self._item_loader.generate_intervals()
+        self._intervals = self._item_loader.generate_intervals(self.subsample_interval)
```
Comment (tchaton) [inline src/litdata/streaming/config.py:63]: I don't think we need to pass it down to the item loader. I think it can just takes a list of chunks and intervals and remap them

File: src/litdata/streaming/dataset.py [hunk 3]
```diff
@@ -412,6 +433,30 @@ def _validate_state_dict(self) -> None:
                 f"Found `{self.drop_last}` instead of `{state['drop_last']}`."
             )
 
+    def _modify_subsample_interval(self, frac_start: float, frac_end: float):
```
Comment (tchaton) [inline src/litdata/streaming/dataset.py:436]: This should be a function, so it is easier to test.

File: src/litdata/streaming/item_loader.py [hunk 4]
```diff
@@ -85,13 +87,20 @@ class PyTreeLoader(BaseItemLoader):
     def __init__(self) -> None:
         self._chunk_filepaths: Dict[str, bool] = {}
 
-    def generate_intervals(self) -> List[Tuple[int, int]]:
+    def generate_intervals(
+        self,
+        subsample_interval: Optional[List[Tuple[int, int]]] = None,
```
Comment (tchaton) [inline src/litdata/streaming/item_loader.py:92]: As I said before. I don't think the item loader should know the re-sampling

File: src/litdata/streaming/shuffle.py [hunk 5]
```diff
@@ -38,14 +38,14 @@ def get_len(self, distributed_env: _DistributedEnv, num_workers: int, batch_size
 
         if self.drop_last:
             items_per_process = [
-                sum((interval[-1] - interval[0]) for interval in intervals) for intervals in intervals_per_ranks
+                sum((interval[2] - interval[1]) for interval in intervals) for intervals in intervals_per_ranks
```
Comment (tchaton) [inline src/litdata/streaming/shuffle.py:41]: This most likely shouldn't change. As below.

File: src/litdata/utilities/train_test_split.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        streaming_dataset (StreamingDataset): An instance of StreamingDataset that needs to be split.
+        splits (List[float]): List of floats representing the proportion of data to be allocated to each split
+                            (e.g., [0.8, 0.1, 0.1] for train, test, and validation).
+
+    Returns:
+        List[StreamingDataset]: A list of StreamingDataset instances, each corresponding to the proportions specified
+                                in the splits argument.
+
+    """
+    if any([not isinstance(split, float) for split in splits]) or not (
```
Comment (tchaton) [inline src/litdata/utilities/train_test_split.py:20]: Let's split the ifs, so we can provide better error message to the users.

File: src/litdata/utilities/train_test_split.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        all([_f > 0 and _f <= 1 for _f in splits]) and sum(splits) <= 1
+    ):
+        raise ValueError("Split should be float with each value in [0,1] and max sum can be 1.")
+
+    my_datasets = [deepcopy(streaming_dataset) for _ in splits]
+
+    frac_start = 0
+    frac_end = 0
```
Comment (tchaton) [inline src/litdata/utilities/train_test_split.py:28]: [0] + np.cumsum(splits) would do the same ;)

**Comment by tchaton** [comment]
I recommend you to always start with writing test when doing development new features. This helps to ensure the changes you are making are correct.

File: src/litdata/utilities/train_test_split.py (hunk 6 above)
Comment (deependujha) [reply src/litdata/utilities/train_test_split.py]: done.
